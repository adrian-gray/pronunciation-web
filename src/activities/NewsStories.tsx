import React, { Component } from "react";

import SplitHilite from "./../components/SplitHilite";
import MemberGate from "./../components/MemberGate";
import Picker from "./../components/Picker";

interface INewsStoriesCustomProps {
  level: string;
  title: string;
  headline: string;
  sentences: string[];
  options: string[];
  answers: string[];
  isUserAuth: boolean;
}

interface INewsStoriesCustomState {
  sentences: any[];
  selectedOption: number[];
  selectedBgColour: string[];
  other?: any[];
}

const countSelectors = (arr: string[]) => {
  let count = 0;
  arr.forEach((sentence: string) => {
    const num = (sentence.match(/OPTION/g) || []).length;
    count += num;
  });
  return count;
};

class NewsStories extends Component<
  INewsStoriesCustomProps,
  INewsStoriesCustomState
> {
  constructor(props: INewsStoriesCustomProps) {
    super(props);

    this.state = {
      sentences: [],
      selectedOption: [],
      selectedBgColour: []
    } as INewsStoriesCustomState;

    this.handleClick = this.handleClick.bind(this);
    this.parseSentences = this.parseSentences.bind(this);
    this.addNewSentences = this.addNewSentences.bind(this);
    this.updateSentences = this.updateSentences.bind(this);
    this.checkForCorrect = this.checkForCorrect.bind(this);
  }

  componentDidMount() {
    this.addNewSentences(this.props.sentences);
  }

  componentWillReceiveProps(nextProps: INewsStoriesCustomProps) {
    if (nextProps.sentences !== this.props.sentences) {
      this.addNewSentences(nextProps.sentences);
    }
  }

  parseSentences(
    sentences: string[],
    selectedOption: number[],
    selectedBgColour: string[]
  ) {
    let id = 0;
    let selectorId = 0;

    const parsedSentences = sentences.map((sentence, index) => {
      const fragments = [];
      let key = 0;
      do {
        const idx = sentence.indexOf("OPTION");
        if (idx === -1) {
          fragments.push(sentence);
          break;
        } else if (idx === 0) {
          fragments.push(
            <Picker
              options={this.props.options}
              selected={selectedOption[selectorId]}
              colour={selectedBgColour[selectorId]}
              handleClick={this.handleClick}
              key={key}
              index={id}
            />
          );
          id++;
          selectorId++;
          sentence = sentence.substr(6, sentence.length);
        } else {
          const substr = sentence.slice(0, idx);
          fragments.push(<SplitHilite str={substr} key={key} />);
          sentence = sentence.slice(substr.length);
        }
        key++;
      } while (sentence.length);

      return <p key={index}>{fragments}</p>;
    });

    return parsedSentences;
  }

  addNewSentences(newSentences: string[]) {
    const numSelectors: number = countSelectors(newSentences);
    const selectedOption: number[] = Array(numSelectors).fill(0);
    const selectedBgColour: string[] = Array(numSelectors).fill(undefined);
    const sentences = this.parseSentences(
      newSentences,
      selectedOption,
      selectedBgColour
    );

    this.setState({
      selectedOption,
      selectedBgColour,
      sentences
    } as INewsStoriesCustomState);
  }

  updateSentences(params: INewsStoriesCustomState) {
    const newSentences = params.sentences;
    const selectedOption = params.selectedOption || this.state.selectedOption;
    const selectedBgColour =
      params.selectedBgColour || this.state.selectedBgColour;

    const sentences = this.parseSentences(
      newSentences,
      selectedOption,
      selectedBgColour
    );

    this.setState({ sentences });
  }

  checkForCorrect(values: number[]) {
    return values.map((value, index: number) => {
      return this.props.options[value] === this.props.answers[index];
    });
  }

  handleClick(index: number) {
    const selectedOption: number[] = this.state.selectedOption.slice();
    const selectedBgColour: string[] = this.state.selectedBgColour.slice();
    selectedOption[index] =
      (this.state.selectedOption[index] + 1) % this.props.options.length;
    if (selectedOption.every(val => val > 0)) {
      this.checkForCorrect(selectedOption).map(
        (value: boolean, index: number) => {
          switch (value) {
            case true:
              selectedBgColour[index] = "correct-bg";
              break;
            default:
              selectedBgColour[index] = "incorrect-bg";
          }
        }
      );
    }
    this.setState({ selectedOption, selectedBgColour });
    this.updateSentences({
      sentences: this.props.sentences,
      selectedOption,
      selectedBgColour
    });
  }

  render() {
    const { headline, title } = this.props;
    const description = `Extra! Extra! Read all about it! News Stories is a collection of interesting and feel-good stories from around the world.  Each story is written in two levels â€“ beginners and intermediate. All you need to do is identify the highlighted sounds and choose the correct phoneme that represents that sound. News Stories helps you to identify individual sounds in long and more difficult written text. This improves your pronunciation of the words you read and your reading fluency.`;
    const content = <div>{this.state.sentences}</div>;

    return (
      <div className="headspace">
        <h3>{"News Stories"}</h3>
        <p>{description}</p>
        <h4>{headline}</h4>
        <h5>
          <SplitHilite str={title} />
        </h5>
        <MemberGate content={content} isUserAuth={this.props.isUserAuth} />
      </div>
    );
  }
}

export default NewsStories;
