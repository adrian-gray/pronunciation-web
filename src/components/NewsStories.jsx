import React, { Component } from "react";

import Paper from "@material-ui/core/Paper";
import Typography from "@material-ui/core/Typography";
import { makeStyles } from "@material-ui/core/styles";
import { withTheme } from "@material-ui/styles";

import SplitHilite from "./SplitHilite";
import MemberGate from "./MemberGate";
import Picker from "./Picker";

const useStyles = makeStyles(theme => ({
  headspace: theme.headspace,
  sentence: {
    padding: "0.5rem",
    paddingLeft: "1rem",
    paddingRight: "1rem",
    fontSize: "1.2rem"
  },
  red: {
    color: "maroon"
  },
  green: {
    color: "green"
  }
}));

const countSelectors = arr => {
  let count = 0;
  arr.forEach(sentence => {
    const num = (sentence.match(/OPTION/g) || []).length;
    count += num;
  });
  return count;
};

function NewsStories(props) {
  const classes = useStyles(props);

  let sentences = [];
  const selectedOption = [];
  let selectedBgColour = [];

  addNewSentences(props.sentences);

  function parseSentences(sentences, selectedOption, selectedBgColour) {
    let id = 0;
    let selectorId = 0;

    const parsedSentences = sentences.map((sentence, index) => {
      const fragments = [];
      let key = 0;
      do {
        const idx = sentence.indexOf("OPTION");
        if (idx === -1) {
          fragments.push(sentence);
          break;
        } else if (idx === 0) {
          fragments.push(
            <Picker
              options={props.options}
              selected={selectedOption[selectorId]}
              colour={selectedBgColour[selectorId]}
              handleClick={handleClick}
              key={key}
              index={id}
            />
          );
          id++;
          selectorId++;
          sentence = sentence.substr(6, sentence.length);
        } else {
          const substr = sentence.slice(0, idx);
          fragments.push(<SplitHilite str={substr} key={key} />);
          sentence = sentence.slice(substr.length);
        }
        key++;
      } while (sentence.length);

      return (
        <Typography key={index} className={classes.sentence}>
          {fragments}
        </Typography>
      );
    });

    return parsedSentences;
  }

  function addNewSentences(newSentences) {
    const numSelectors = countSelectors(newSentences);
    const selectedOption = Array(numSelectors).fill(0);
    selectedBgColour = Array(numSelectors).fill(undefined);
    sentences = parseSentences(newSentences, selectedOption, selectedBgColour);
  }

  function updateSentences(params) {
    const newSentences = params.sentences;
    const selectedOption = params.selectedOption || selectedOption;
    selectedBgColour = params.selectedBgColour || selectedBgColour;

    sentences = parseSentences(newSentences, selectedOption, selectedBgColour);
  }

  function checkForCorrect(values) {
    return values.map((value, index) => {
      return props.options[value] === props.answers[index];
    });
  }

  function handleClick(index) {
    if (!props.userAuth) return;
    const newStateSelectedOption = selectedOption.slice();
    newStateSelectedOption[index] =
      (selectedOption[index] + 1) % props.options.length;
    if (newStateSelectedOption.every(val => val > 0)) {
      checkForCorrect(newStateSelectedOption).map((value, index) => {
        switch (value) {
          case true:
            selectedBgColour[index] = classes.green;
            break;
          case false:
            selectedBgColour[index] = classes.red;
            break;
          default:
            selectedBgColour[index] = undefined;
        }
      });
    }
    tsetState({
      selectedOption: newStateSelectedOption,
      selectedBgColour
    });
    updateSentences({
      sentences: props.sentences,
      selectedOption: newStateSelectedOption,
      selectedBgColour
    });
  }

  const { headline, title } = props;
  const content = <Paper>{sentences}</Paper>;
  const description = `Extra! Extra! Read all about it! News Stories is a collection of interesting and feel-good stories from around the world.  Each story is written in two levels â€“ beginners and intermediate. All you need to do is identify the highlighted sounds and choose the correct phoneme that represents that sound. News Stories helps you to identify individual sounds in long and more difficult written text. This improves your pronunciation of the words you read and your reading fluency.`;

  return (
    <div className={classes.headspace}>
      <Typography variant="h5" gutterBottom>
        {"News Stories"}
      </Typography>
      <Typography variant="body2" gutterBottom>
        {description}
      </Typography>
      <Typography variant="h5" gutterBottom>
        {headline}
      </Typography>
      <Typography variant="h6" gutterBottom>
        <SplitHilite str={title} />
      </Typography>
      <MemberGate content={content} userAuth={props.userAuth} />
    </div>
  );
}

export default withTheme(NewsStories);
