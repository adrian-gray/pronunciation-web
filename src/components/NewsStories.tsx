import React, { Component } from "react";

import Container from "react-bootstrap/Container";

import SplitHilite from "./SplitHilite";
import MemberGate from "./MemberGate";
import Picker from "./Picker";

const countSelectors = arr => {
  let count = 0;
  arr.forEach(sentence => {
    const num = (sentence.match(/OPTION/g) || []).length;
    count += num;
  });
  return count;
};

interface NewsStoriesCustomProps {
  title: string;
  headline: string;
  sentences: string[];
  options: string[];
  selectOption: string;
  answers: string[];
  userAuth: number;
  other: any[];
}

interface NewsStoriesCustomState {
  sentences: string[];
  selectedOption: number[];
  selectedBgColour: string[];
  other: any[];
}

class NewsStories extends Component<
  NewsStoriesCustomProps,
  NewsStoriesCustomState
> {
  constructor(props) {
    super(props);

    const { ...other } = props;

    this.state = {
      sentences: [],
      selectedOption: [],
      selectedBgColour: [],
      other: []
    };

    this.handleClick = this.handleClick.bind(this);
    this.parseSentences = this.parseSentences.bind(this);
    this.addNewSentences = this.addNewSentences.bind(this);
    this.updateSentences = this.updateSentences.bind(this);
    this.checkForCorrect = this.checkForCorrect.bind(this);
  }

  componentDidMount() {
    this.addNewSentences(this.props.sentences);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.sentences !== this.props.sentences) {
      this.addNewSentences(nextProps.sentences);
    }
  }

  parseSentences(sentences, selectedOption, selectedBgColour) {
    let id = 0;
    let selectorId = 0;

    const parsedSentences = sentences.map((sentence, index) => {
      const fragments = [];
      let key = 0;
      do {
        const idx = sentence.indexOf("OPTION");
        if (idx === -1) {
          fragments.push(sentence);
          break;
        } else if (idx === 0) {
          fragments.push(
            <Picker
              options={this.props.options}
              selected={selectedOption[selectorId]}
              colour={selectedBgColour[selectorId]}
              handleClick={this.handleClick}
              key={key}
              index={id}
              {...this.state.other}
            />
          );
          id++;
          selectorId++;
          sentence = sentence.substr(6, sentence.length);
        } else {
          const substr = sentence.slice(0, idx);
          fragments.push(<SplitHilite str={substr} key={key} />);
          sentence = sentence.slice(substr.length);
        }
        key++;
      } while (sentence.length);

      return <p key={index}>{fragments}</p>;
    });

    return parsedSentences;
  }

  addNewSentences(newSentences) {
    const numSelectors = countSelectors(newSentences);
    const selectedOption = Array(numSelectors).fill(0);
    const selectedBgColour = Array(numSelectors).fill(undefined);
    const sentences = this.parseSentences(
      newSentences,
      selectedOption,
      selectedBgColour
    );

    this.setState({
      selectedOption,
      selectedBgColour,
      sentences
    });
  }

  updateSentences(params) {
    const newSentences = params.sentences;
    const selectedOption = params.selectedOption || this.state.selectedOption;
    const selectedBgColour =
      params.selectedBgColour || this.state.selectedBgColour;

    const sentences = this.parseSentences(
      newSentences,
      selectedOption,
      selectedBgColour
    );

    this.setState({ sentences });
  }

  checkForCorrect(values) {
    return values.map((value, index) => {
      return this.props.options[value] === this.props.answers[index];
    });
  }

  handleClick(index) {
    if (!this.props.userAuth) return;
    const newStateSelectedOption = this.state.selectedOption.slice();
    const selectedBgColour = [];
    newStateSelectedOption[index] =
      (this.state.selectedOption[index] + 1) % this.props.options.length;
    if (newStateSelectedOption.every(val => val > 0)) {
      this.checkForCorrect(newStateSelectedOption).map((value, index) => {
        switch (value) {
          case true:
            selectedBgColour[index] = "#99BB99"; // TODO
            break;
          case false:
            selectedBgColour[index] = "#BB9999"; // TODO
            break;
          default:
            selectedBgColour[index] = undefined;
        }
      });
    }
    this.setState({
      selectedOption: newStateSelectedOption,
      selectedBgColour
    });
    this.updateSentences({
      sentences: this.props.sentences,
      selectedOption: newStateSelectedOption,
      selectedBgColour
    });
  }

  render() {
    const { headline, title } = this.props;

    const content = <Container>{this.state.sentences}</Container>;

    const description = `Extra! Extra! Read all about it! News Stories is a collection of interesting and feel-good stories from around the world.  Each story is written in two levels â€“ beginners and intermediate. All you need to do is identify the highlighted sounds and choose the correct phoneme that represents that sound. News Stories helps you to identify individual sounds in long and more difficult written text. This improves your pronunciation of the words you read and your reading fluency.`;

    return (
      <Container className="headspace">
        <h2>{"News Stories"}</h2>
        <p>{description}</p>
        <h3>{headline}</h3>
        <h5>
          <SplitHilite str={title} />
        </h5>
        <MemberGate
          content={content}
          userAuth={this.props.userAuth}
          {...this.props.other}
        />
      </Container>
    );
  }
}

export default NewsStories;
