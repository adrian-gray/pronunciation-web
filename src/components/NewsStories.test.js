import React from "react";
import { shallow } from "enzyme";
import { render, cleanup } from "@testing-library/react";

import NewsStories from "./NewsStories";

afterEach(cleanup);

test("NewsStories", async () => {
  const props = {};

  const container = shallow(<NewsStories {...props} />);

  expect(container.firstChild).toMatchSnapshot();
});
