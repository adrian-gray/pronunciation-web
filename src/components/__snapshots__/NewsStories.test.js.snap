// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsStories 1`] = `undefined`;
